server/src/main/java/org/elasticsearch/action/support/FanOutListener.java

```diff
@@ -0,0 +1,294 @@
+    }
+
+    @Override
+    public final void onFailure(Exception exception) {
+        setResult(new FailureResult(exception, wrapException(exception)));
+    }
+
+    protected Exception wrapException(Exception exception) {
+        return exception;
+    }
+
+    /**
+     * @return {@code true} if and only if this listener has been completed (either successfully or exceptionally).
+     */
+    public final boolean isDone() {
+        return isDone(ref.get());
+    }
+
+    /**
+     * @return the result with which this listener completed successfully, or throw the exception with which it failed.
+     *
+     * @throws AssertionError if this listener is not complete yet and assertions are enabled.
+     * @throws IllegalStateException if this listener is not complete yet and assertions are disabled.
+     */
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    public final T rawResult() throws Exception {
+        final var refValue = ref.get();
+        if (refValue instanceof SuccessResult result) {
+            return (T) result.result();
+        } else if (refValue instanceof FailureResult result) {
+            throw result.exception();
+        } else {
+            assert false : "not done";
+            throw new IllegalStateException("listener is not done, cannot get result yet");
+        }
+    }
+
+    protected static RuntimeException wrapAsExecutionException(Throwable t) {
+        if (t instanceof RuntimeException runtimeException) {
+            return runtimeException;
+        } else {
+            return new UncategorizedExecutionException("Failed execution", new ExecutionException(t));
+        }
+    }
+
+    private static <T> ActionListener<T> preserveContext(@Nullable ThreadContext threadContext, ActionListener<T> listener) {
+        return threadContext == null ? listener : ContextPreservingActionListener.wrapPreservingContext(listener, threadContext);
+    }
+
+    private static <T> ActionListener<T> fork(Executor executor, ActionListener<T> listener) {
+        return executor == EsExecutors.DIRECT_EXECUTOR_SERVICE ? listener : new ThreadedActionListener<>(executor, listener);
+    }
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    private static <T> boolean tryComplete(Object refValue, ActionListener<T> listener) {
+        if (refValue instanceof SuccessResult result) {
+            result.complete(listener);
+            return true;
+        }
+        if (refValue instanceof FailureResult result) {
+            result.complete(listener);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * If incomplete, atomically update {@link #ref} with the given result and use it to complete any pending listeners.
+     */
+    @SuppressWarnings("unchecked")
+    private void setResult(Object result) {
+        assert isDone(result);
+
+        var currentValue = ref.get();
+        while (true) {
+            if (isDone(currentValue)) {
```

Comment:
In what situations would more than one thing try to set result at the same time?